src/modules/python/pbs/v1/_pmi_cray.py

```diff
@@ -0,0 +1,424 @@
+#
+#  Altair’s dual-license business model allows companies, individuals, and
+#  organizations to create proprietary derivative works of PBS Pro and distribute
+#  them - whether embedded or bundled with other software - under a commercial
+#  license agreement.
+#
+#  Use of Altair’s trademarks, including but not limited to "PBS™",
+#  "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+#  trademark licensing policies.
+ *
+ */
+"""
+__doc__ = """
+This module is used for Cray systems.
+"""
+
+import pbs
+import os
+import sys
+import stat
+from pbs.v1._pmi_types import BackendError
+from pbs.v1._pmi_utils import _running_excl, _pbs_conf, _get_vnode_names, \
+    _svr_vnode
+
+pbsexec = _pbs_conf("PBS_EXEC")
+if pbsexec is None:
+    raise BackendError("PBS_EXEC not found")
+
+sys.path.append(os.path.join(pbsexec, "python", "lib", "python2.7"))
+sys.path.append(os.path.join(pbsexec, "python", "lib", "python2.7",
+                "lib-dynload"))
+
+
+def launch(id, args):
+    """
+    Run capmc and return the structured output.
+    """
+    import json
+
+    # full path to capmc given by Cray
+    cmd = "/opt/cray/capmc/default/bin/capmc"
+    if not os.path.exists(cmd):
+        cmd = "capmc"		# should be in PATH then
+    cmd = cmd + " " + args
+    fail = ""
+
+    pbs.logjobmsg(id, "launch: " + cmd)
+    stdin, stdout, stderr = os.popen3(cmd)
+    stdin.close()
+
+    try:
+        out = json.load(stdout)
+    except:
+        out = None
+    stdout.close()
+
+    err = stderr.readline()		# first line only
+    stderr.close()
+
+    pid, stat = os.wait()
+    if os.WIFEXITED(stat):
+        exitval = os.WEXITSTATUS(stat)
+        if exitval != 0:
+            fail = "%s: exit %d" % (cmd, exitval)
+        else:
+            pbs.logjobmsg(id, "launch: finished")
+    else:
+        fail = "%s: no exit status" % cmd
+
+    if out is not None:
+        errno = out["e"]
+        msg = out["err_msg"]
+        if errno != 0 or (len(msg) > 0 and msg != "Success"):
+            fail = "output: e=%d err_msg='%s'" % (errno, msg)
+    if len(err) > 0:
+        pbs.logjobmsg(id, "stderr: %s" % err.strip())
+
+    if len(fail) > 0:
+        pbs.logjobmsg(id, fail)
+        raise BackendError(fail)
+    return out
+
+
+def jobnids(job):
+    # return the set of nids belonging to a job
+
+    nidset = set()
+    craynid = "PBScraynid"
+    for vname in _get_vnode_names(job):
+        vnode = _svr_vnode(vname)
+        try:
+            nidset.add(int(vnode.resources_available[craynid]))
+        except:
+            pass
+    return nidset
+
+
+def nodenids(hosts):
+    # return the set of nids from the host list
+    nidset = set()
+    craynid = "PBScraynid"
+    for vnames in hosts:
+        vnode = _svr_vnode(vnames)
+        try:
+            nidset.add(int(vnode.resources_available[craynid]))
+        except:
+            pass
+    return nidset
+
+
```

Comment:
Describe the function and its arguments as per pep8 guidelines. Please take care of this wherever required.

Reply:
Done.